assets/js/googlesitekit/data/create-notifications-store.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * Provides API functions to create a datastore for notifications.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+import { keyBy, omit } from 'lodash';
+import { get } from 'googlesitekit-api';
+
+const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
+const REMOVE_NOTIFICATION = 'REMOVE_NOTIFICATION';
+const FETCH_NOTIFICATIONS = 'FETCH_NOTIFICATIONS';
+const RECEIVE_NOTIFICATIONS = 'RECEIVE_NOTIFICATIONS';
+const RECEIVE_NOTIFICATIONS_FAILED = 'RECEIVE_NOTIFICATIONS_FAILED';
+
+// This should remain private for now, hence not be exported on Data.
+export const createNotificationsStore = ( type, identifier, datapoint ) => {
+	const INITIAL_STATE = {
+		serverNotifications: undefined,
+		clientNotifications: {},
+		isFetchingNotifications: false,
+	};
+
+	const actions = {
+		addNotification( notification ) {
+			return {
+				notification,
+				type: ADD_NOTIFICATION,
+			};
+		},
+
+		removeNotification( id ) {
+			return {
+				id,
+				type: REMOVE_NOTIFICATION,
+			};
+		},
+
+		fetchNotifications() {
+			return {
+				payload: {},
+				type: FETCH_NOTIFICATIONS,
+			};
+		},
+
+		receiveNotifications( notifications ) {
+			invariant( notifications, 'notifications is required.' );
+
+			return {
+				payload: { notifications },
+				type: RECEIVE_NOTIFICATIONS,
+			};
+		},
+
+		receiveNotificationsFailed( error ) {
+			invariant( error, 'error is required.' );
+
+			return {
+				payload: {},
+				type: RECEIVE_NOTIFICATIONS_FAILED,
+			};
+		},
+	};
+
+	const controls = {
+		[ FETCH_NOTIFICATIONS ]: () => {
+			return get( type, identifier, datapoint );
+		},
+	};
+
+	const reducer = ( state = INITIAL_STATE, action ) => {
+		switch ( action.type ) {
+			case ADD_NOTIFICATION: {
+				const { notification } = action;
+
+				return {
+					...state,
+					clientNotifications: {
+						...state.clientNotifications,
+						...keyBy( [ notification ], 'id' ),
+					},
+				};
+			}
+
+			case REMOVE_NOTIFICATION: {
+				const { id } = action;
+
+				return {
+					...state,
+					clientNotifications: omit( state.clientNotifications, [ id ] ),
```

Comment:
I'd rather this be an array and have this return `state.clientNotifications.filter((notification) => id !== notification.id )`.

If you want to keep this as an object it would be better/faster to copy the object and delete the relevant key. `omit` is  slower as it goes through the object (and it's actually being removed in the next version of lodash: https://github.com/lodash/lodash/issues/2930).

The benefit to this being an object would be the ability to delete a notification based on its key without iterating across all the notifications actually, so maybe a good reason to store them that way:

```js
case REMOVE_NOTIFICATION: {
		const { id } = action;

                const newNotifications = { ...state.clientNotifications };
                delete newNotifications[id];

		return {
			...state,
			clientNotifications: newNotifications,
```

Less lodash and more standard ES6 😄

Reply:
Actually, my bad here, `clientNotifications` already is an object. I just shouldn't be using `omit` here.

Just to confirm, if you use `delete` on something which doesn't exist/is undefined, that will just silently do nothing rather than throw an error right?

Reply:
Yup! (Bizarrely enough, it will actually return `true`. JS is weird. 😄)

<img width="286" alt="Screenshot 2020-03-09 at 23 20 21" src="https://user-images.githubusercontent.com/90871/76265502-8f5d6f80-625c-11ea-8ee2-33b4843f8c2a.png">